---
title: "dasatinib_dif_exp_analysis_29_11_19"
author: "Kieran Redpath"
date: "29 November 2019"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

NOTE: Don't run this at the same time as "dasatinib_dif_exp_analysis_3_tissues", cause I'm lazy and didn't change the names of objects

limma differential expression analysis (lecture 3)
look at high and low ic50 groups, and then find out which genes are differentially expressed between the two groups
data frame with all the samples that have dasatinib data, and all the genes. These will have high/low ic50 associated with them (50/50 split)
Will show you which genes are involved in dasatinib sensitivity
fit model with lmfit and ebayes, gives you a table with p values, showing significance of differential expression
volcano plot tells you how big and significant the difference is, stuff that comes out the top will be the most important
Pathway analysis to see what it's actually changing

```{r Load packages}
library(limma)
library(edgeR)
library(data.table)
library(magrittr)
library(ggplot2)
library(CePa)
library(tidyverse)
library(ggbeeswarm)
library(org.Hs.eg.db)
library(dplyr)
```



```{r Create the first big file to use, results='hide'}
# Output suppressed cause setDT prints a whole lot of useless stuff
# expDat Processing
ccleExpData <- read.gct('CCLE_GDSC/Data/CCLE_RNAseq_genes_counts_20180929.gct')
logExpDat <- log(ccleExpData + 0.5)
dge <- DGEList(counts=ccleExpData)
dge <- calcNormFactors(dge)
v <- voom(dge, plot=TRUE)
expDat <- v$E

# GDSC2 Processing (from old document)
GDSC2 <- fread('CCLE_GDSC/Data/GDSC2_fitted_dose_response_15Oct19.csv', sep= ';')
dim(GDSC2)
dasatinib <- GDSC2 %>% filter(., DRUG_NAME=="Dasatinib")
expDatsplit <- strsplit(colnames(expDat), "_") %>% 
  lapply(., function(x) x[1]) %>% 
  unlist()
colnames(expDat) <- expDatsplit
dasatinib$CELL_LINE_NAME <- gsub("-","",dasatinib$CELL_LINE_NAME, fixed=TRUE)
commonSamples <- intersect(dasatinib$CELL_LINE_NAME,colnames(expDat))
expDat_match <- match(commonSamples, colnames(expDat))
expDat_sort <- expDat[ , na.omit(expDat_match)]
dasatinib_match <- match(commonSamples, dasatinib$CELL_LINE_NAME)
dasatinib_sort <- dasatinib[na.omit(dasatinib_match),]
colnames(dasatinib_sort)[5] <- "CCLE_Name"
dasatinib_sort$AUC_Level <- ifelse(dasatinib_sort$AUC < quantile(dasatinib_sort$AUC, 0.5, na.rm=TRUE), "Low","High")
# expDat Processing
expDat_all <- t(expDat_sort)
expDat_all <- as.data.frame(expDat_all)
setDT(expDat_all, keep.rownames = "CCLE_Name")[]
# "dasatinib_sort" can now be combined with "expDat_sort"
dasatinib_all = full_join(x=dasatinib_sort, y= expDat_all, by=c("CCLE_Name" = "CCLE_Name"))
```



```{r Differential Expression Analysis With IC50}
# Check all this stuff
dim(expDat_sort)
dim(dasatinib_sort)
head(expDat_sort)[,1:5]
head(dasatinib_sort)[,1:5]
sum(colnames(expDat_sort)==dasatinib_sort$CCLE_Name)
names(dasatinib_sort)

# lmFit
group <- ifelse(dasatinib_sort$LN_IC50 > median(dasatinib_sort$LN_IC50), "High", "Low")
table(group)
boxplot(dasatinib_sort$LN_IC50 ~ group)
design = model.matrix(~group);
design %>% head()
colnames(design) = c("Mean"
,"HighVsLow"
)

# Fit the data to the model
fit = lmFit(expDat_sort, design)
fit = eBayes(fit)
tt = topTable(fit, coef="HighVsLow", adjust="BH",n=nrow(expDat_sort))
options(digits=4)
tt[1:20,]
# Take a quick look at the first gene for the hell of it
plot(density(expDat_sort[,1]))

# Count number of significant genes
sum(tt$adj.P.Val<0.01)

## Plot log fold-change _versus_ -log(P-value). This graph is the important one, things up the top and spread out are useful associations
## (i.e., higher number = lower p-value):
volcanoplot(fit, coef="HighVsLow")

split <- strsplit(rownames(tt),".", fixed=T) %>% lapply(., function(x) x[1]) %>% unlist()
geneNames <- AnnotationDbi::select(org.Hs.eg.db, keys = split, column = c("SYMBOL","GENENAME"), key="ENSEMBL")
head(split)

# What does the top table look like?
dim(tt)
tt$symbol <- geneNames$SYMBOL[match(split, geneNames$ENSEMBL)]
setDT(tt, keep.rownames = TRUE)[]

# Check the data: Output shows that some ENSEMBL ID's have multiple regular gene names
length(split)
dim(geneNames)

# Print the top 300 associated genes. Can then look them up on Enrichr, GeneSetDB, etc
cat(na.omit(tt$symbol[1:300]),sep="\n")
#Check log fold change
sign(tt$logFC)
# Create a histogram of IC50 values to visualise them
hist(dasatinib_sort$LN_IC50,50)
# Take a quick look at the median, just out of interest
median(dasatinib_sort$LN_IC50)
head(tt)

#Find the top expression values. Shows that high expression correlates to high ic50, and V.V (e.g. this gene is protective against dasatinib)
topExp <- expDat_sort[match(tt$rn[1], rownames(expDat_sort)),]
df <- data.frame(topGene=topExp, ic50=group)
ggplot(df, aes(x=ic50, y=topGene)) + geom_boxplot()
```



```{r Differential Expression Analysis With AUC}
# Check all this stuff
dim(expDat_sort)
dim(dasatinib_sort)
head(expDat_sort)[,1:5]
head(dasatinib_sort)[,1:5]
sum(colnames(expDat_sort)==dasatinib_sort$CCLE_Name)
names(dasatinib_sort)

# lmFit
group2 <- ifelse(dasatinib_sort$AUC > median(dasatinib_sort$AUC), "High", "Low")
table(group2)
boxplot(dasatinib_sort$AUC ~ group2)
design2 = model.matrix(~group2);
design2 %>% head()
colnames(design2) = c("Mean"
,"HighVsLow"
)
# Fit the data to the model
fit2 = lmFit(expDat_sort, design2)
fit2 = eBayes(fit2)
tt2 = topTable(fit2, coef="HighVsLow", adjust="BH",n=nrow(expDat_sort))
options(digits=4)
tt2[1:20,]
# Take a quick look at the first gene for the hell of it
plot(density(expDat_sort[,1]))

# Count number of significant genes
sum(tt2$adj.P.Val<0.01)

## Plot log fold-change _versus_ -log(P-value). This graph is the important one, things up the top and spread out are useful associations
## (i.e., higher number = lower p-value):
volcanoplot(fit2, coef="HighVsLow")

split2 <- strsplit(rownames(tt2),".", fixed=T) %>% lapply(., function(x) x[1]) %>% unlist()
geneNames2 <- AnnotationDbi::select(org.Hs.eg.db, keys = split2, column = c("SYMBOL","GENENAME"), key="ENSEMBL")
head(split2)

# What does the top table look like?
dim(tt2)
tt2$symbol <- geneNames2$SYMBOL[match(split2, geneNames2$ENSEMBL)]
setDT(tt2, keep.rownames = TRUE)[]

# Check the data: Output shows that some ENSEMBL ID's have multiple regular gene names
length(split2)
dim(geneNames2)

# Print the top 300 associated genes. Can then look them up on Enrichr, GeneSetDB, etc
cat(na.omit(tt2$symbol[1:300]),sep="\n")
# Create a histogram of AUC values to visualise them
hist(dasatinib_sort$AUC,50)
# Take a quick look at the median, just out of interest
median(dasatinib_sort$AUC)
head(tt2)

#Find the top expression values. Shows that high expression correlates to low AUC, and V.V (e.g. this gene is protective against dasatinib)
topExp2 <- expDat_sort[match(tt2$rn[1], rownames(expDat_sort)),]
df2 <- data.frame(topGene=topExp2, AUC=group)
ggplot(df2, aes(x=AUC, y=topGene)) + geom_boxplot()
```



```{r Find the overlap between the significant samples}
# Create a table with ensembl ID, gene symbol, adj. p value, and sign for AUC and IC50
tt3 <- full_join(tt, tt2, by= "rn")
tt3 <- dplyr::select(tt3, -c("symbol.x", "P.Value.y", "P.Value.x", "t.y", "t.x", "B.x", "B.y", "AveExpr.x"))
tt3 <- dplyr::rename(tt3, "Gene_Symbol" = "symbol.y", "AUC_logFC" = "logFC.y", "Avg_Exp" = "AveExpr.y", "AUC_Adj_PVal" = "adj.P.Val.y", "IC50_logFC" = "logFC.x", "IC50_Adj_PVal" = "adj.P.Val.x", "Ensembl_ID" = "rn")
tt3 <- tt3[c(7,1,5,2,3,4,6)]
# Find out if the log fold change sign is consistent with significance. Remember, -logFC relates to resistance
tt3 <- mutate(tt3, sign(tt3$IC50_logFC), sign(tt3$AUC_logFC))
tt3 <- mutate(tt3, sign(tt3$IC50_logFC)*sign(tt3$AUC_logFC))
# Multiply signs together, keep genes where the result is negative (cause you want two different signs)
tt3 <- dplyr::rename(tt3, "logFC_Sign" = "sign(tt3$IC50_logFC) * sign(tt3$AUC_logFC)", "IC50_Sign" = "sign(tt3$IC50_logFC)", "AUC_Sign" = "sign(tt3$AUC_logFC)")
# Look at the samples that are potentially important- don't get good results :(
###################### IMPORTANT!!! ######################
### IF ALL THE DATA IS ACTUALLY -LOG, NOT LN, THEN THE RELATIONSHIP BETWEEN AUC AND IC50(-log) SHOULD BE +VE. THIS MEANS THAT THE SIGN WE'RE LOOKING FOR WILL BE +VE, NOT -VE! FROM HERE, WE'RE ASSUMING THEIR LABELLING IS WRONG! Adding in +ve AUC sign now for SigSamples, genes that are related to susceptibility to dasatinib will have a larger AUC (and theoretically larger "-log" IC50)
# Coming back to this, 05/12/19
# Filter based on significance, increased susceptibility to dasatinib, larger log fold change and positive correlation between AUC and "-log" IC50
SigSamples <- filter(tt3, tt3$logFC_Sign == 1, tt3$IC50_Adj_PVal < 0.01, tt3$AUC_Adj_PVal < 0.01, abs(tt3$IC50_logFC) > 1, abs(tt3$AUC_logFC) > 1)
# Rank the samples based on average rank between the two p-values
SigSamples <- 
  SigSamples %>% mutate(., rank_ic50=rank(IC50_Adj_PVal)) %>%
  mutate(., rank_auc=rank(AUC_Adj_PVal)) %>% 
  mutate(., avg_rank=0.5*(rank_auc + rank_ic50)) %>% 
  arrange(., avg_rank)
View(SigSamples)
# Find the top 100 ranked genes, and look em up
cat(na.omit(SigSamples$Gene_Symbol[1:400]),sep="\n")
# Sort based on average significance rank across AUC and IC50 (filter based on significant adj. p values, and consistent signs between the two analysis)
tt3 <- 
  tt3 %>% mutate(., rank_ic50=rank(IC50_Adj_PVal)) %>%
  mutate(., rank_auc=rank(AUC_Adj_PVal)) %>% 
  mutate(., avg_rank=0.5*(rank_auc + rank_ic50)) %>% 
  arrange(., avg_rank)

View(tt3)
# When you actually get a good list of genes, analyse with enrichr and genesetDB
```



```{r Plot gene expression vs AUC and gene expression vs IC50, on a scatter plot and boxplot}
# Gene expression vs IC50
# Boxplot
ggplot(data=tt3, mapping=aes(x=IC50_Sign,y=Avg_Exp, group=IC50_Sign, fill=IC50_Sign)) +
  geom_boxplot()
ggplot(data=tt3, mapping=aes(x=IC50_logFC,y=Avg_Exp)) +geom_point()

# Gene expression vs AUC
# Boxplot
ggplot(data=tt3, mapping=aes(x=AUC_Sign,y=Avg_Exp, group=AUC_Sign, fill=AUC_Sign)) +
  geom_boxplot()
# Scatter plot
ggplot(data=tt3, mapping=aes(x=AUC_logFC,y=Avg_Exp)) +geom_point()
```

Looks like either it's all -logged, or the relationship between ic50 and auc is not what we think it is.

### Next Steps:

### Find cell lines that are allegedly resistant/susceptible to a drug, and figure out what their IC50's are- are they high or low when they should be high or low (i.e resistant/susceptible, not neccesarily in that order) Should look up breast cancer cell lines, probably drugs like paclitaxel, doxarubicin, etc.

### Paclitaxel resistant breast cancer cell lines have higher paclitaxel *non log* ic50 values (e.g. 62nM 5 day) than non-resistant cells (e.g. 5 nM 5 day). This looks like it associates with lower AUC: On a dose-response curve, the y axis has % inhibition, and the x axis has concentration of the drug. so, if there's increased IC50, the concentration at 50% inhibition is further to the right of the graph, and there's less underneath the curve.

### Then go back to looking at signs (+/-, "dasatinib_dif_exp_analysis_29_11_19"), and trying to figure out what's associated with different ic50/auc. It might be different to what you think, depending on if your initial assumptions were based on regular or -logged data. From here, find a good list of significantly associated genes.

Do a volcano plot of your significant genes again maybe???
Filter your final list based on fold change. -ve or +ve???
Work through the stuff in slides 18-36 with your own data.

Look up the paper on GOseq. GOseq does pathway analysis, adjusted by gene length (which is quite important), which is available as an r package. This allows you to use r to carry out similar processes as enrichr, but with adjusted data.
ReactomePA is also available, but doesn't adjust for gene length.

Also look at your significant gene list on enrichr.



```{r ReactomePA Analysis from Mik's Slides, fig.height=3, fig.width=8}
# Load packages
library(ReactomePA)
library(org.Hs.eg.db)

# Create list of top associated genes
Top400Genes <- as.character(na.omit(SigSamples$Gene_Symbol[1:400]))
# Add Entrez ID's
hs <- org.Hs.eg.db
Top400GenesEntrez <- select(hs, 
                            keys = Top400Genes,
                            columns = c("ENTREZID", "SYMBOL"),
                            keytype = "SYMBOL")

# Perform pathway analysis via ReactomePA
# Gives you a nice table with some useful pathways
dasat_rPAoverrep <- enrichPathway(gene=Top400GenesEntrez$ENTREZID, organism = "human",
 pvalueCutoff=0.05, readable=T)
View(as.data.frame(dasat_rPAoverrep))
# Visualise these as plots (note that they look weird in the inline output, have to be opened properly)
barplot(dasat_rPAoverrep, showCategory=11, font.size=8)
dotplot(dasat_rPAoverrep, showCategory=11, font.size=8)
```



```{r GoSeq Analysis}
## Load packages
library(reactome.db)
library(goseq)
library(gplots)
## Set random set
set.seed(321)

# Look at the existing file that you'll use again here
View(Top400GenesEntrez)

## Get pathway names, and extract human-only pathways
rName <- as.list(reactomePATHNAME2ID)
rName <- rName[grep("Homo sapiens", names(rName))]

## List of all reactome pathways, and entrez IDs of the genes they contain
rGenes <- as.list(reactomePATHID2EXTID)
## Extract ONLY the human pathways 
rGenesPath <- rGenes[match(rName, names(rGenes))]
## Make sure that each gene is only listed once per pathway
rGenesPath <- lapply(rGenesPath, unique)
## Get list of genes, and the pathways they are in.
rGeneByPath <- as.list(reactomeEXTID2PATHID)
## myGenes is the list of all the genes in your data set (e.g., the rownames of 
## your expression matrix): Entrez IDs
## allGenes is the intersection of the genes from the data set with the human reactome genes
allGenes <- intersect( Top400GenesEntrez$ENTREZID, unique(unlist(rGenesPath)) )
length(allGenes)

####### Unsure what's going  on from here #######
## sigGenes is the binary list of whether or not the gene was significant.
## same length and order as allGenes
## Here I've just randomly generated 0's and 1's. SHOULD I NOT BE DOING THIS?
sigGenes <- rep(0, length(allGenes))
names(sigGenes) <- allGenes
## Randomly generate significant genes, based on gene length (longer transcripts have higher chance of being significant)
geneLengths <- getlength(allGenes, "hg19", id="knownGene")

randSig <- which( runif(length(allGenes)) < geneLengths/20000 )
sigGenes[randSig] <- 1
sum(sigGenes)

## Figure out which genes are relevant to your data set (some may not be present)
mt <- match(allGenes, names(rGeneByPath))
## Count how many of the reactome genes aren't in your data set
sum(is.na(mt))

## Get genes and pathways relevant to our analysis
rGeneByPath <- lapply(rGeneByPath[mt], function(x) intersect(x, names(rGenesPath)))
head(rGeneByPath)

## GOseq analysis
pwf <- nullp(sigGenes, 'hg19', id = "knownGene", plot.fit=TRUE)

## Length-corrected
goseqReactome <- goseq(pwf, gene2cat = rGeneByPath)

## Non-length-corrected
hyperReactome <- goseq(pwf, gene2cat = rGeneByPath, method="Hypergeometric")

## Adjust p-values
goseqReactome$adjP <- p.adjust(goseqReactome$over_represented_pvalue, method="fdr")
hyperReactome$adjP <- p.adjust(hyperReactome$over_represented_pvalue, method="fdr")

## Check results
head(goseqReactome)
head(hyperReactome)

## Using 0.2 as FDR significance threshold, just to show impact of not correcting.
## Remember, these are randomly significant genes, but the chance of being significant still depends on their length.
venn(list(goseq=goseqReactome$category[goseqReactome$adjP < 0.2],
          hyper=hyperReactome$category[hyperReactome$adjP < 0.2]))
```


